cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go

```diff
@@ -458,3 +459,132 @@ func TestValidateFeatureGates(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateArgSelection(t *testing.T) {
+	var tests = []struct {
+		name     string
+		c        *kubeadm.NodeConfiguration
+		expected bool
+	}{
+		{
+			"invalid: DiscoveryToken and DiscoveryFile cannot both be set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:  "foo",
+				DiscoveryToken: "abcdef.1234567890123456@foobar",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryToken or DiscoveryFile must be set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:  "",
+				DiscoveryToken: "",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenAPIServers not set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken: "abcdef.1234567890123456@foobar",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "foo",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "foo",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "foo",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
+		},
+		{
+			"WARNING: using token-based discovery without DiscoveryTokenCACertHashes can be unsafe",
+			&kubeadm.NodeConfiguration{
+				DiscoveryTokenUnsafeSkipCAVerification: false,
+			},
+			false,
+		},
+
+		{
+			"WARNING: kubeadm doesn't fully support multiple API Servers yet",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:           "abcdef.1234567890123456@foobar",
+				DiscoveryTokenAPIServers: []string{"192.168.122.100:9898", "192.168.122.88:9898"},
+			},
+			true,
+		},
```

Comment:
Should we have more valid scenarios here ? Token + Servers + CACertHashes ?

Reply:
Maybe I wanted to add valid such valid scenarios but I forgot those copy&paste things :) Will address it.